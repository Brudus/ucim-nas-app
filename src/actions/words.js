import database from '../firebase/firebase';

// ADD_WORD
export const addWord = (word) => ({
    type: 'ADD_WORD',
    word
});

export const startAddWord = (wordData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            source = '', 
            destination= '', 
            repeatAt = 0,
            interval = 0,
            reps = 0,
            easeFactor = 2.5,
            repeatAtInverted = 0,
            intervalInverted = 0,
            repsInverted = 0,
            easeFactorInverted = 2.5
        } = wordData;
        const word = { source, destination, repeatAt, interval, reps, easeFactor, repeatAtInverted, intervalInverted, repsInverted, easeFactorInverted };
        
        return database.ref(`users/${uid}/words`).push(word).then((ref) => {
            dispatch(addWord({
                id: ref.key,
                ...word
            }));
        });
    };
};

// REMOVE_WORD
export const removeWord = ({ id } = {}) => ({
    type: 'REMOVE_WORD',
    id
});

export const startRemoveWord = ({ id } = {}) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words/${id}`).remove().then(() => {
            dispatch(removeWord({ id }));
        });
    };
};

// EDIT_WORD
export const editWord = (id, updates) => ({
    type: 'EDIT_WORD',
    id,
    updates
});

export const startEditWord = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words/${id}`).update(updates).then(() => {
            dispatch(editWord(id, updates));
        })
    };
};

// SET_WORDS
export const setWords = (words) => ({
    type: 'SET_WORDS',
    words
});

export const startSetWords = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words`).once('value').then((snapshot) => {
            const words = [];

            snapshot.forEach((childSnapshot) => {
                words.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            
            dispatch(setWords(words));
        });
    };
};