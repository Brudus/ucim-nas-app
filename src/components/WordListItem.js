import React from 'react';
import { Link } from 'react-router-dom';

const formatDate = (dateToFormat) => {
    const nowDate = Date.now();
    const dateDifference = dateToFormat - nowDate;
    const dayInMs = 1000 * 60 * 60 * 24;

    if (dateDifference <= 0) {
        return 'now';
    } else if (dateDifference < dayInMs) {
        return 'today';
    } else if (dateDifference < 2 * dayInMs) {
        return 'tomorrow';
    } else {
        return `in ${Math.floor(dateDifference / dayInMs)} days`;
    }
};

const WordListItem = ({ id, source, destination, repeatAt, repeatAtInverted, isNew, isNewInverted }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{destination}</h3>
        </Link>
        <h4>{source}</h4>
        {(isNew || isNewInverted) && <span>NEW</span>}
        <p>German -> Serbian - Next repetition: {formatDate(repeatAt)}</p>
        <p>Serbian -> German - Next repetition: {formatDate(repeatAtInverted)}</p>
    </div>
);

export default WordListItem;
    
    