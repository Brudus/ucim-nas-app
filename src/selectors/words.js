export default (words, { text, sortBy }) => {
    return words.filter((word) => {
        const sourceText = word.source.toLowerCase()
        const destinationText = word.destination.toLowerCase();
        const searchText = text.toLowerCase().trim();

        const sourceMatch = sourceText.includes(searchText);
        const destinationMatch = destinationText.includes(searchText);
        return sourceMatch || destinationMatch;
    }).sort((a, b) => {
        if (sortBy === 'repeatAt') {
            return a.repeatAt > b.repeatAt ? 1 : -1;
        } else if (sortBy === 'source') {
            const aText = a.source.toLowerCase().trim();
            const bText = b.source.toLowerCase().trim();
            return aText > bText ? 1 : -1;
        } else if (sortBy === 'destination') {
            const aText = a.destination.toLowerCase().trim();
            const bText = b.destination.toLowerCase().trim();
            return aText > bText ? 1 : -1;
        }
    });
};