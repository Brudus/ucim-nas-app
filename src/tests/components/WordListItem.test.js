import React from 'react';
import { shallow } from 'enzyme';
import WordListItem from '../../components/WordListItem';
import words from '../fixtures/words';

test('should render WordListItem with word (now)', () => {
    const wrapper = shallow(<WordListItem key={words[0].id} {...words[0]} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render WordListItem correctly with word (today)', () => {
    const halfADay = 0.5 * 1000 * 60 * 60 * 24;
    const repeatAt = Date.now() + halfADay;
    const changedWord = {...words[1], repeatAt};
    const wrapper = shallow(<WordListItem key={words[1].id} {...changedWord} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render WordListItem correctly with word (tomorrow)', () => {
    const oneAndAHalfDay = 1.5 * 1000 * 60 * 60 * 24;
    const repeatAt = Date.now() + oneAndAHalfDay;
    const changedWord = {...words[1], repeatAt};
    const wrapper = shallow(<WordListItem key={words[1].id} {...changedWord} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render WordListItem correctly with word (in 2 days)', () => {
    const twoAndAHalfADay = 2.5 * 1000 * 60 * 60 * 24;
    const repeatAt = Date.now() + twoAndAHalfADay;
    const changedWord = {...words[1], repeatAt};
    const wrapper = shallow(<WordListItem key={words[1].id} {...changedWord} />);
    expect(wrapper).toMatchSnapshot();
});
