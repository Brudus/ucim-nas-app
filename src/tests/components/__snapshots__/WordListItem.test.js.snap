// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render WordListItem correctly with word (in 2 days) 1`] = `
<div>
  <Link
    to="/edit/2"
  >
    <h3>
      nisam
    </h3>
  </Link>
  <h4>
    bin ich nicht
  </h4>
  <p>
    Next repetition: 
    in 2 days
  </p>
</div>
`;

exports[`should render WordListItem correctly with word (today) 1`] = `
<div>
  <Link
    to="/edit/2"
  >
    <h3>
      nisam
    </h3>
  </Link>
  <h4>
    bin ich nicht
  </h4>
  <p>
    Next repetition: 
    today
  </p>
</div>
`;

exports[`should render WordListItem correctly with word (tomorrow) 1`] = `
<div>
  <Link
    to="/edit/2"
  >
    <h3>
      nisam
    </h3>
  </Link>
  <h4>
    bin ich nicht
  </h4>
  <p>
    Next repetition: 
    tomorrow
  </p>
</div>
`;

exports[`should render WordListItem with word (now) 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      dobro
    </h3>
  </Link>
  <h4>
    gut
  </h4>
  <p>
    Next repetition: 
    now
  </p>
</div>
`;
